src/python/pants/core/goals/lint.py

```diff
@@ -225,61 +228,81 @@ async def lint(
+        "Iterable[type[LintFilesRequest]]", union_membership[LintFilesRequest]
+    )
 
-    _check_ambiguous_request_names(*target_request_types, *file_request_types)
+    _check_ambiguous_request_names(
+        *lint_target_request_types, *fmt_target_request_types, *file_request_types
+    )
 
     specified_names = determine_specified_tool_names(
         "lint",
         lint_subsystem.only,
-        target_request_types,
+        [
+            *lint_target_request_types,
+            *fmt_target_request_types,
+        ],
         extra_valid_names={request.name for request in file_request_types},
     )
-    target_requests = tuple(
-        request_type(
-            request_type.field_set_type.create(target)
-            for target in targets
-            if (
-                request_type.name in specified_names
-                and request_type.field_set_type.is_applicable(target)
-            )
-        )
-        for request_type in target_request_types
-    )
 
-    def address_str(fs: FieldSet) -> str:
-        return fs.address.spec
+    def is_specified(request_type: type[StyleRequest] | type[LintFilesRequest]):
+        return request_type.name in specified_names
+
+    lint_target_request_types = filter(is_specified, lint_target_request_types)
+    fmt_target_request_types = filter(is_specified, fmt_target_request_types)
+    file_request_types = filter(is_specified, file_request_types)
 
     def batch(field_sets: Iterable[FieldSet]) -> Iterator[list[FieldSet]]:
-        return partition_sequentially(
+        partitions = partition_sequentially(
             field_sets,
-            key=address_str,
+            key=lambda fs: fs.address.spec,
             size_target=lint_subsystem.batch_size,
             size_max=4 * lint_subsystem.batch_size,
         )
+        for partition in partitions:
+            yield partition
+
+    def batch_by_type(
+        request_types: Iterable[type[_SR]],
+    ) -> tuple[tuple[type[_SR], list[FieldSet]], ...]:
+        return tuple(
+            (request_type, field_set_batch)
+            for request_type in request_types
+            for field_set_batch in batch(
+                request_type.field_set_type.create(target)
+                for target in targets
+                if request_type.field_set_type.is_applicable(target)
+            )
+        )
 
     lint_target_requests = (
-        request.__class__(field_set_batch)
-        for request in target_requests
-        if isinstance(request, LintTargetsRequest) and request.field_sets
-        for field_set_batch in batch(request.field_sets)
+        request_type(batch) for request_type, batch in batch_by_type(lint_target_request_types)
+    )
+
+    batched_fmt_request_pairs = batch_by_type(fmt_target_request_types)
+    all_fmt_source_batches = await MultiGet(
+        Get(
+            SourceFiles,
+            SourceFilesRequest(
+                getattr(field_set, "sources", getattr(field_set, "source", None))
```

Comment:
This makes me nervous. Could we add validation to the fieldset that these are defined?